torch/csrc/jit/operator.cpp

```diff
@@ -203,49 +207,179 @@ struct SchemaParser {
+  // Once numbers are always treated as seperate types from Tensors, this line
+  // should be removed, since it opens up the possibility of ambigous declarations
+  // dispatching to the wrong implementation.
+  if ((formal->isSubtypeOf(*NumberType::get()) ||
+       formal->isSubtypeOf(*ListType::ofInts())) &&
+      actual->isSubtypeOf(*DynamicType::get()))
+    return true;
+
+  return false;
+}
+
+bool Operator::matchesNode(Node* node) const {
+  size_t attributes_size = node->numAttributes();
+  size_t attributes_seen = 0;
+  auto inputs_size = node->inputs().size();
+  size_t input_i = 0;
+  for(size_t arg_i = 0; arg_i < schema.arguments.size(); ++arg_i) {
+    at::optional<AttributeKind> attribute_kind;
+    const Argument& arg = schema.arguments[arg_i];
+    if(attributes_size > 0 && (attribute_kind = attributeKindOf(arg.type))) {
+      auto name = Symbol::fromQualString("attr::" + arg.name);
+      if(!node->hasAttribute(name) || node->kindOf(name) != *attribute_kind) {
+        // std::cout << "missing attribute: " << name << "\n";
+        return false;
+      }
+      attributes_seen++;
+    } else if(*arg.type == *ListType::ofTensors()) {
+      // Tensor[] is handled as varargs, consume inputs until the remaining required arguments
+      // XXX - there can only be a single Tensor[] in a declaration
+      size_t remaining_required = 0;
+      for(size_t j = arg_i + 1; j < schema.arguments.size(); ++j){
+        // remaining arguments are only those that won't be consumed from attributes
+        if(attributes_size == 0 || !attributeKindOf(schema.arguments[j].type))
+          remaining_required++;
+      }
+      while(inputs_size - input_i > remaining_required) {
+        auto input = node->inputs()[input_i++];
+        if(!typeMatches(input->type(), DynamicType::get())) {
+          // std::cout << "vararg argument is not Dynamic\n";
+          return false;
+        }
+      }
+    } else {
+      if(input_i == inputs_size) {
+        // std::cout << "not enough inputs\n";
+        return false;
+      }
+      auto input = node->inputs()[input_i++];
+      if(!typeMatches(input->type(), arg.type)) {
+        // std::cout << "argument " << arg_i << " has the wrong type\n";
+        return false;
+      }
+    }
+  }
+
+  if(!schema.is_vararg && input_i != inputs_size) {
+    // std::cout << "not all inputs used\n" << input_i << " " << inputs_size << "\n";
+    return false;
+  }
+  if(!schema.is_vararg && attributes_seen != attributes_size) {
+    // std::cout << "not all attributes used\n" << attributes_seen << " " << attributes_size << "\n";
+    return false;
+  }
+  return true;
+}
+
+std::shared_ptr<Operator> findOperatorFor(Node* node) {
+  const auto& candidates = getOperatorsFor(node->kind());
+  for(const auto& candidate : candidates) {
+    if(candidate->matchesNode(node)) {
+      return candidate;
+    }
+  }
+  return nullptr;
+}
+
+const Operator& getOperatorFor(Node* node) {
```

Comment:
The fact that we have getOperatorFor call findOperatorFor, which in turns calls getOperatorsFor is a bit funny (it initially looked like infinite recursion to me)